### PR TITLE
fix(validation): parse CVSS vector-only audit severities

### LINKED ISSUE #1679: Follow-up from #1424: 1 item(s) (critical_bug)
_Consolidated follow-up from implementation of #1424._

Each section below lists scope-checked follow-up items discovered during implementation. Items are restricted to core library defects, security, safety hazards, or critical functional bugs.

### CVSS vector-only severity entries silently dropped in audit.py

hephaestus/validation/audit.py:77-86 — extract_cvss_score() matches a CVSS vector string (e.g. "CVSS:3.1/AV:N/AC:L/...") at line 77 but deliberately skips it (pass, line 81), recovering a numeric score only from entry['base_score']/entry['cvss_score'] at line 82. A pip-audit severity entry that provides ONLY a vector string and no separate base_score/cvss_score field yields no score; the vuln then falls through severity_label()->'UNKNOWN' and is classed non-blocking in filter_audit_results() (line 147), letting a real HIGH/CRITICAL finding bypass the CI gate. Repro: feed extract_cvss_score([{'score': 'CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H'}]) -> returns None (should derive 9.8). Sketch fix: parse the base score out of the vector (the metrics map to a computable base score, or at minimum extract a leading 'CVSS:x.y' is insufficient — implement the CVSS v3 base-score formula or pull the numeric score pip-audit attaches), instead of pass. Add a unit test in tests/unit/validation/test_audit.py asserting a vector-only entry produces the correct numeric severity.

---

_The implementer also considered the items below and rejected them as out of scope; they are recorded in the PR body, not filed as separate issues._

### PR DESCRIPTION
**Pull request by mvillmow**
## Summary
Fixes audit severity extraction so CVSS vector-only pip-audit entries produce numeric scores instead of falling through as UNKNOWN and being silently suppressed.

## Changes
- `hephaestus/validation/audit.py`
  - `extract_cvss_score` now scores CVSS v3.0/3.1 vector strings found in `score` (in addition to plain numeric values), so vector-only HIGH/CRITICAL findings stay blocking.
  - Fixed the falsy-zero bug: `base_score`/`cvss_score` are each parsed independently instead of `entry.get("base_score") or entry.get("cvss_score")`, which discarded legitimate `0`/`"0"` values and masked `cvss_score` behind a falsy `base_score`.
  - `_parse_cvss_numeric_score` rejects NaN/inf, values outside the inclusive 0.0-10.0 CVSS range, and `bool` inputs.
  - `_cvss_v3_impact`: removed the incorrect CVSS:3.0 branch that applied the v3.1 Environmental ModifiedImpact form (`ISS*0.9731`, exponent 13). The base-score changed-scope Impact formula is identical in CVSS 3.0 and 3.1: `7.52*(ISS-0.029) - 3.25*(ISS-0.02)^15`. The unused `version` parameter was dropped.
- `tests/unit/validation/test_audit.py`
  - Vector-only scoring: `CVSS:3.1/.../S:U/C:H/I:H/A:H` = 9.8, changed-scope 3.1 and 3.0 vectors = 9.9 (official values).
  - HIGH-boundary regression test: `CVSS:3.0/AV:A/AC:L/PR:H/UI:R/S:C/C:H/I:L/A:L` scores 7.1 per the base spec (ISS = 0.732304; the removed branch yielded 7.0).
  - Lowercase vector normalization, malformed-vector rejection, out-of-range/NaN/bool numeric-score rejection, and vector-only blocking behavior in `filter_audit_results`.

## Testing
```
$ pixi run pytest tests/unit/validation/test_audit.py -q
39 passed in 0.20s
```
Also green: `pixi run pytest tests/unit/validation -q` (872 passed, 16 skipped), `pixi run mypy` (no issues in 493 files), `pixi run ruff check hephaestus/ tests/`, `pre-commit run --all-files`.

Closes #1679

Generated by Codex via ProjectHephaestus automation.

### CODE REVIEW

File: tests/unit/validation/test_audit.py [hunk 1]
```diff
@@ -56,9 +56,19 @@ def test_string_numeric_score(self) -> None:
...
-        """Returns None when no numeric score is available."""
+    def test_string_score_does_not_skip_base_score(self) -> None:
+        """Considers base_score even when score is a numeric string."""
+        result = extract_cvss_score([{"score": "5.0", "base_score": "9.1"}])
+        assert result == 9.1
+
+    def test_cvss_vector_score(self) -> None:
+        """Computes the base score from a CVSS v3 vector-only entry."""
         result = extract_cvss_score([{"score": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H"}])
+        assert result == 9.8
```
Comment (mvillmow) [inline tests/unit/validation/test_audit.py:72]: This regression covers the unchanged-scope CVSS v3.1 path, but the implementation also adds separate changed-scope scoring logic. Add a changed-scope vector case, for example `CVSS:3.1/AV:N/AC:L/PR:L/UI:N/S:C/C:H/I:H/A:H` -> `9.9`, so that high-risk formula branch is guarded.

**Review by mvillmow — COMMENTED** [review]
GO; vector-only CVSS findings are now scored and blocking, with one minor test coverage gap.

File: hephaestus/validation/audit.py [hunk 2]
```diff
@@ -68,16 +80,80 @@ def extract_cvss_score(severity_list: list[dict[str, Any]]) -> float | None:
...
     for entry in severity_list:
-        score_str = entry.get("score", "")
-        if isinstance(score_str, (int, float)):
-            scores.append(float(score_str))
+        score = entry.get("score", "")
+        if isinstance(score, (int, float)):
+            scores.append(float(score))
+        elif isinstance(score, str):
+            with contextlib.suppress(ValueError):
+                scores.append(float(score))
```
Comment (mvillmow) [inline hephaestus/validation/audit.py]: `float(score)` accepts non-finite strings such as `nan` and `inf`. With `{"score": "nan", "base_score": "9.8"}`, this appends `nan` before the real base score, `max([nan, 9.8])` returns `nan`, and `severity_label(nan)` falls through to `NONE`, suppressing a blocking vulnerability. Parse into a temporary value and append it only when it is finite and in the CVSS range, e.g. `math.isfinite(parsed) and 0.0 <= parsed <= 10.0`; apply the same guard anywhere string scores are converted.

**Review by mvillmow — COMMENTED** [review]
NOGO: new string score parsing can turn malformed non-finite values into audit-gate bypasses.

File: hephaestus/validation/audit.py [hunk 3]
```diff
@@ -68,16 +92,83 @@ def extract_cvss_score(severity_list: list[dict[str, Any]]) -> float | None:
...
+    if scope == "C":
+        raw_score *= 1.08
+    return _cvss_v3_round_up(min(raw_score, 10.0))
+
+
+def _cvss_v3_impact(version: str, scope: str, impact_subscore: float) -> float:
+    if scope == "U":
+        return 6.42 * impact_subscore
+    if version == "CVSS:3.0":
+        return 7.52 * (impact_subscore - 0.029) - 3.25 * (impact_subscore * 0.9731 - 0.02) ** 13
```
Comment (mvillmow) [inline hephaestus/validation/audit.py]: This uses the wrong formula for CVSS:3.0 Scope Changed base impact. The v3.0 base equation is `7.52 * (ISS - 0.029) - 3.25 * (ISS - 0.02) ** 15`; the `* 0.9731` / `** 13` form belongs to modified-impact scoring, not v3.0 Base. As written, `CVSS:3.0/AV:P/AC:H/PR:L/UI:N/S:C/C:H/I:H/A:L` scores 6.9 here but should round to 7.0, so a vector-only HIGH finding is suppressed. Use the v3.0 base formula and add a boundary regression test.

**Review by mvillmow — COMMENTED** [review]
NOGO: CVSS:3.0 changed-scope vectors can be under-scored below the HIGH blocking threshold.

File: hephaestus/validation/audit.py [hunk 4]
```diff
@@ -68,16 +92,83 @@ def extract_cvss_score(severity_list: list[dict[str, Any]]) -> float | None:
...
+        for field in ("base_score", "cvss_score"):
+            parsed = _parse_cvss_numeric_score(entry.get(field))
+            if parsed is not None:
+                scores.append(parsed)
     return max(scores) if scores else None
 
 
+def _score_cvss_v3_vector(vector: str) -> float | None:
+    stripped = vector.strip().upper()
+    if not stripped.startswith(("CVSS:3.0/", "CVSS:3.1/")):
```
Comment (mvillmow) [inline hephaestus/validation/audit.py:119]: This accepts CVSS:3.0 vectors, but the scoring path uses the CVSS v3.1 Scope Changed impact formula. CVSS v3.0 uses the older changed-scope term, so boundary vectors can cross the audit gate incorrectly: `CVSS:3.0/AV:P/AC:H/PR:L/UI:N/S:C/C:H/I:H/A:L` should score 6.9, but this implementation returns 7.0 and blocks it as HIGH. Either branch the impact formula by vector version or stop accepting CVSS:3.0 vectors.

**Review by mvillmow — COMMENTED** [review]
NOGO: CVSS:3.0 vectors are accepted but scored with the v3.1 changed-scope formula, causing boundary misclassification.

File: hephaestus/validation/audit.py [hunk 5]
```diff
@@ -68,16 +92,84 @@ def extract_cvss_score(severity_list: list[dict[str, Any]]) -> float | None:
...
+            if vector_score is not None:
+                scores.append(vector_score)
+        for field in ("base_score", "cvss_score"):
+            parsed = _parse_cvss_numeric_score(entry.get(field))
+            if parsed is not None:
+                scores.append(parsed)
     return max(scores) if scores else None
 
 
+def _score_cvss_v3_vector(vector: str) -> float | None:
```
Comment (mvillmow) [inline hephaestus/validation/audit.py:117]: `_score_cvss_v3_vector` has no docstring. The metric-table lookup structure and CVSS-spec formula are non-obvious to a future reader; a one-line summary (e.g. `"""Compute the CVSS v3 base score from a vector string, or None if malformed."""`) would match the style of every other function in this file and make the private helper self-documenting without adding verbosity.

**Review by mvillmow — COMMENTED** [review]
Correct CVSS v3 base-score implementation with comprehensive tests; one private helper missing a docstring.

File: hephaestus/validation/audit.py [hunk 6]
```diff
@@ -68,16 +92,84 @@ def extract_cvss_score(severity_list: list[dict[str, Any]]) -> float | None:
...
+        return 0.0
+
+    raw_score = impact + exploitability
+    if scope == "C":
+        raw_score *= 1.08
+    return _cvss_v3_round_up(min(raw_score, 10.0))
+
+
+def _cvss_v3_impact(scope: str, impact_subscore: float) -> float:
+    # The changed-scope Impact formula is identical in CVSS 3.0 and 3.1
```
Comment (mvillmow) [inline hephaestus/validation/audit.py:164]: `_cvss_v3_impact` accepts `scope: str` but is only defined for `"U"` and `"C"` — any other value silently falls through to the changed-scope formula and returns a wrong result. The caller guards scope at line 132, so this is benign today, but consider asserting or returning `None` for unexpected values to avoid silent misbehavior if called from future code paths.

**Review by mvillmow — COMMENTED** [review]
Core bug fixed correctly; CVSS v3 formula spec-accurate; tests comprehensive. One minor POLA gap in private helper; no blockers.

File: hephaestus/validation/audit.py [hunk 7]
```diff
@@ -68,16 +92,84 @@ def extract_cvss_score(severity_list: list[dict[str, Any]]) -> float | None:
...
+        score = entry.get("score", "")
+        if isinstance(score, (int, float)):
+            parsed = _parse_cvss_numeric_score(score)
+            if parsed is not None:
+                scores.append(parsed)
+        elif isinstance(score, str):
+            parsed = _parse_cvss_numeric_score(score)
+            if parsed is not None:
+                scores.append(parsed)
+            vector_score = _score_cvss_v3_vector(score)
```
Comment (mvillmow) [inline hephaestus/validation/audit.py]: A numeric string score (e.g. `"5.0"`) passes through `_parse_cvss_numeric_score` successfully and then also reaches `_score_cvss_v3_vector`, which immediately returns `None` because `"5.0"` doesn't start with `"CVSS:3.0/"` or `"CVSS:3.1/"`. Functionally harmless, but a micro-optimization would be to only call `_score_cvss_v3_vector` when `_parse_cvss_numeric_score` returned `None` (i.e. the string failed to parse as a plain number). No behavior change required.

**Review by mvillmow — COMMENTED** [review]
Correct CVSS v3 base-score formula; vector-only bypass fixed; tests proportional and cover key boundary cases. One minor redundant call.

File: hephaestus/validation/audit.py (hunk 7 above)
Comment (mvillmow) [reply hephaestus/validation/audit.py]: Fixed in 86db329: vector scoring now runs only when numeric parsing failed (a numeric string can never also be a CVSS vector). 39 audit tests green.

**Comment by mvillmow** [comment]
**Operator GO decision** (same #1554-churn pattern as PR #1837): reviewer verdict was GO on substance in all three rounds (R0/R1/R2), stranding only on one minor thread per round at the iteration cap; the final thread (a self-described micro-optimization) is fixed in 86db329 and resolved. Strict full-coverage audit: CONDITIONAL→fixed (CVSS 3.0 changed-scope formula corrected against the base spec, hand-verified vectors 9.9/7.1, 39 tests green, CI green). Applying implementation-go and arming squash auto-merge per policy.
